### PR TITLE
Refactoring of compiler pipe line

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

**Review by henriknyman — APPROVED** [review]

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/CypherCompiler.scala [hunk 1]
```diff
@@ -19,135 +19,37 @@
...
                           cacheMonitor: CypherCacheFlushingMonitor[CacheAccessor[Statement, ExecutionPlan]],
-                          monitors: Monitors) {
+                          monitors: Monitors,
+                          sequencer: String => RewriterStepSequencer) {
 
   def planQuery(queryText: String, context: PlanContext, notificationLogger: InternalNotificationLogger,
                 plannerName: String = "",
                 offset: Option[InputPosition] = None): (ExecutionPlan, Map[String, Any]) = {
-    planPreparedQuery(prepareSyntacticQuery(queryText, queryText, notificationLogger, plannerName), notificationLogger, context, offset, CompilationPhaseTracer.NO_TRACING)
+    val step1: State4 = prepareSyntacticQuery(queryText, queryText, notificationLogger, plannerName)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/CypherCompiler.scala]: `step1` is a `State4`? Perhaps `step1State` or `state` is a more appropriate name?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/CypherCompiler.scala [hunk 2]
```diff
@@ -157,66 +59,37 @@ case class CypherCompiler(parser: CypherParser,
...
-    val rewrittenSyntacticQuery = syntacticQuery.rewrite(rewriteProcedureCalls(context.procedureSignature, context.functionSignature))
+  def prepareSemanticQuery(in: State4,
+                           notificationLogger: InternalNotificationLogger,
+                           context: PlanContext,
+                           tracer: CompilationPhaseTracer = CompilationPhaseTracer.NO_TRACING): State4 = {
+    val rewriteProcedureCalls = new RewriteProcedureCalls(context.procedureSignature, context.functionSignature)
 
-    val procedureDeprecations = procedureDeprecationNotifications(rewrittenSyntacticQuery.statement)
-    procedureDeprecations.foreach(notificationLogger.log)
+    val secondPipeLine =
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/CypherCompiler.scala]: Any specific reason for this one being method-temporary, but the `firstPipeline` being object-temporary (asymmetry)?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/CypherCompilerFactory.scala [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      rewriterSequencer = rewriterSequencer,
+      plannerName = plannerName,
+      runtimeBuilder = runtimeBuilder,
+      config = config,
+      updateStrategy = updateStrategy,
+      publicTypeConverter = typeConverter.asPublicType
+    )
+    val procedurePlanProducer = DelegatingProcedureExecutablePlanBuilder(costPlanProducer, typeConverter.asPublicType)
+
+    // Pick planner based on input
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/CypherCompilerFactory.scala]: This comment looks misplaced. We're picking the planner on line 55, no?

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/phases/CompilationState.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+package org.neo4j.cypher.internal.compiler.v3_2.phases
+
+import org.neo4j.cypher.internal.compiler.v3_2.tracing.rewriters.RewriterCondition
+import org.neo4j.cypher.internal.frontend.v3_2.ast.{Query, Statement}
+import org.neo4j.cypher.internal.frontend.v3_2.{InputPosition, SemanticState, SemanticTable}
+
+object CompilationState {
+
+  /*
+  The reason for all these traits is to make sure that for each stage, we add but do not loose information
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/phases/CompilationState.scala]: lose*

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/CostBasedExecutablePlanBuilder.scala [hunk 5]
```diff
@@ -108,24 +106,16 @@ object CostBasedExecutablePlanBuilder {
...
-    )
 
-    val state = semanticChecker.check(namespacedStatement.toString, namespacedStatement, mkException = (msg, pos) => throw new InternalException(s"Unexpected error during late semantic checking: $msg at $pos"))
+    val state = SemanticChecker.check(statement, mkException = (msg, pos) => throw new InternalException(s"Unexpected error during late semantic checking: $msg at $pos"))
     val table = semanticTable.copy(types = state.typeTable, recordedScopes = state.recordedScopes)
 
-    val newStatement = statementRewriter.rewriteStatement(namespacedStatement)(
+    val newStatement = statementRewriter.rewriteStatement(statement)(
       collapseMultipleInPredicates,
       nameUpdatingClauses /* this is actually needed as a precondition for projectedNamedPaths even though we do not handle updates in Ronja */ ,
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/CostBasedExecutablePlanBuilder.scala]: This comment is obviously out of date :)

**Review by Mats-SX — COMMENTED** [review]
Great improvements!

File: community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/CostBasedExecutablePlanBuilder.scala [hunk 6]
```diff
@@ -44,38 +43,27 @@ case class CostBasedExecutablePlanBuilder(monitors: Monitors,
...
         val (periodicCommit, logicalPlan, pipeBuildContext) = closing(tracer.beginPhase(LOGICAL_PLANNING)) {
-          produceLogicalPlan(ast, rewrittenSemanticTable)(planContext, planContext.notificationLogger())
+          produceLogicalPlan(ast, inputQuery.semanticTable)(planContext, planContext.notificationLogger())
         }
-          runtimeBuilder(periodicCommit, logicalPlan, pipeBuildContext, planContext, tracer, rewrittenSemanticTable,
+          runtimeBuilder(periodicCommit, logicalPlan, pipeBuildContext, planContext, tracer, inputQuery.semanticTable,
                          planBuilderMonitor, plannerName, inputQuery, createFingerprintReference, config)
       case x =>
-        throw new CantHandleQueryException(x.toString())
+        throw new InternalException(s"Can't plan a $x query with the cost planner")
```
Comment (henriknyman) [inline community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/CostBasedExecutablePlanBuilder.scala:66]: Shouldn't we keep the CantHandleQueryException to try fallback on 3.1 rule planner?

**Pull request merged**